File: Core-backend-layer/Programming-fundamentals/Python/OOP/src/abstract_classes/customers/customer.py
from products.product import Product

# Python does not have a modifier to declare a class as final.
class Customer:
    def __init__(self, name):
        self.name = name

    def purchase_product(self, product: Product):
        print(f"{self.name} is purchasing:")
        product.display_info()
        print()

File: Core-backend-layer/Programming-fundamentals/Java/OOP/src/AbstractClasses/customers/Customer.java
package customers;

import products.Product;

// Final class - cannot be subclassed
public final class Customer {
    String name;

    public Customer(String name) {
        this.name = name;
    }

    public void purchaseProduct(Product product) {
        System.out.println(name + " is purchasing:");
        product.displayInfo();
        System.out.println();
    }
}